### DISCUSSION TITLE
concurrent builds and downstream passed resource triggering

### DISCUSSION THREAD

**Discussion by fauna-chase** [concourse/concourse#6327] [Help & Support]
Hi,

I'm currently trying to understand the behavior of concourse when I allow builds to execute concurrently for a job and this job has a downstream job based on a passed constraint.  I'll make use of the following example to illustrate the question:

I have a pipeline with a build job that is triggered by a github repository, downstream from this action I have a deploy job that uses a passed constraint on the github resource from the build job.  2 commits are merged into the tracking branch for the repository in fast succession, commit 1 happening first and commit 2 (the later commit) happening second.  Both commits were merged in at close to the same time so 2 builds for my build job are started.  For some reason, the build for commit 1 runs slower than the build for commit 2, which triggers the deploy job upon completion based on the passed constraint.  

Now, sometime later the build for commit 1 finishes, will this also trigger a build for the deploy job, given that it is an older commit?

Looking at the docs provided here: https://concourse-ci.org/scheduler.html
it seems that it would because the version would be different of that in the previous build, which was commit 2.  In this scenario I wouldn't want a deploy build to be triggered because it is for an older commit.  It seems that I could make my build job serialized to avoid this out of order scenario, but if possible I'd like to allow for parallel execution to increase the rate that changes flowed through.

Let me know if I can provide any more info.

Thanks!
Chase

**Comment by taylorsilva** [comment]
I believe what will happen is that your deploy job will run with commit 2 and never run with commit 1. The scheduler in Concourse should trigger the job with the latest version of the resource that meets its passed constraint.
So when commit 1 eventually finishes it will not trigger your deploy job because your deploy job already ran with a newer version.

The only way this would NOT happen is if your deploy job has [`version: every`](https://concourse-ci.org/jobs.html#schema.step.get-step.version) set. The default for this is setting is `latest`.

**Reply by fauna-chase** [reply]
one thing I'm struggling with in this regard is, is the scenario I described similar to resource pinning?  In the scheduler docs I see

> What this means is if the algorithm runs and computes an input version, the scheduler will create a new build as long as that version is different than the previous build's version for that same input. Even if that version has been used by a build 2 months ago, the scheduler will schedule a new build as long as it has not been used by the previous build.

And in this scenario the build for commit 1 would be different than the later commit.  Is it that the Algorithm mentioned in the Build Scheduler docs would always compute the input version in a passed constraint to be the version produced from the latest started build?

So in a situation where a job kicks of build 1 and then build 2, but build 1 finishes after build 2, the algorithm still computes the version to be passed as the version produced by build 2 because it started after build 1.

**Reply by taylorsilva** [reply]
> So in a situation where a job kicks of build 1 and then build 2, but build 1 finishes after build 2, the algorithm still computes the version to be passed as the version produced by build 2 because it started after build 1.

I believe you are correct on how this scenario would play out. and in this scenario it sounds like you wouldn't care if build 1 was never passed to the job downstream.
